ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/castor/algorithm/CastorFullOptimization.java

```diff
@@ -736,13 +729,124 @@ static void applyPreventiveResultsForAutoOrCurativeRangeActions(Network network,
+        Set<RangeAction<?>> rangeActionsToRemove = multipleInstantRangeActions.stream()
+                .filter(ra -> {
+                    if (ra instanceof PstRangeAction pstRangeAction) {
+                        return pstRangeAction.getRanges().stream().anyMatch(tapRange -> tapRange.getRangeType().equals(RELATIVE_TO_PREVIOUS_INSTANT));
+                    }
+                    return ((StandardRangeAction<?>) ra).getRanges().stream().anyMatch(standardRange -> standardRange.getRangeType().equals(RELATIVE_TO_PREVIOUS_INSTANT));
+                })
+                .collect(Collectors.toSet());
+
+        // Removes RAs that put crac RaUsageLimits at risk.
+        // First, we filter out state that diverged because we know no set-point was chosen for this state.
+        Map<State, OptimizationResult> newContingencyResults = new HashMap<>(contingencyResults);
+        newContingencyResults.entrySet().removeIf(entry -> entry.getValue() instanceof SkippedOptimizationResultImpl);
+
+        // Then, we build a map that gives for each RA, its tap at each state it's available at.
+        Set<RangeAction<?>> rangeActionsNotPreventive = crac.getRangeActions().stream().filter(ra -> isRangeActionAutoOrCurative(ra, crac)).collect(Collectors.toSet());
+        State preventiveState = crac.getPreventiveState();
+        rangeActionsToRemove.forEach(multipleInstantRangeActions::remove);
+        Map<RangeAction<?>, Map<State, Set<Double>>> setPointResults = buildSetPointResultsMap(crac, firstPreventiveResult, newContingencyResults, multipleInstantRangeActions, rangeActionsNotPreventive, preventiveState);
+
+        // Finally, we filter out RAs that put crac RaUsageLimits at risk.
+        rangeActionsToRemove.addAll(getRangeActionsToRemove(crac, preventiveState, setPointResults));
+        return rangeActionsToRemove;
+    }
+
+    /**
+     * Creates a map that gives for a given RA, its tap for each state it's available at.
+     * The only subtlety being that RAs sharing exactly the same network elements are considered to be only one RA.
+     */
+    private static Map<RangeAction<?>, Map<State, Set<Double>>> buildSetPointResultsMap(Crac crac, PerimeterResult firstPreventiveResult, Map<State, OptimizationResult> contingencyResults, Set<RangeAction<?>> multipleInstantRangeActions, Set<RangeAction<?>> rangeActionsNotPreventive, State preventiveState) {
+        Map<RangeAction<?>, Set<RangeAction<?>>> correspondanceMap = new HashMap<>();
+        crac.getRangeActions().stream().filter(ra -> isRangeActionPreventive(ra, crac) && !isRangeActionAutoOrCurative(ra, crac)).forEach(pra -> {
+            Set<NetworkElement> praNetworkElements = pra.getNetworkElements();
+            rangeActionsNotPreventive.forEach(cra -> {
+                if (cra.getNetworkElements().equals(praNetworkElements)) {
+                    correspondanceMap.putIfAbsent(pra, new HashSet<>(Set.of(cra)));
```

Comment:
you can just create an empty HashSet here since you add the cra on the next line